include/kernel.h

```diff
@@ -749,10 +756,9 @@ extern void k_thread_foreach_unlocked(
  * K_FP_REGS, and K_SSE_REGS. Multiple options may be specified by separating
  * them using "|" (the logical OR operator).
  *
- * Historically, users often would use the beginning of the stack memory region
- * to store the struct k_thread data, although corruption will occur if the
- * stack overflows this region and stack protection features may not detect this
- * situation.
+ * The stack_size parameter must be the same size value used when the stack
+ * object was defined, or the return value of K_THREAD_STACK_SIZEOF() on the
+ * stack object.
```

Comment:
This is very nice @andrewboie , it really clears it up. I have a suggestion to make it more explicit:
(and, btw, K_THREAD_STACK_SIZEOF is now renamed to K_KERNEL_STACK_SIZEOF.)

```
The stack_size parameter must be the same size value used when the stack
object was defined using K_KERNEL_STACK_DEFINE(), or, equivalently, the
return value of K_KERNEL_STACK_SIZEOF() on the stack object.
```

So, to summarize:
- `STACKSIZE` (value in bytes) shall **always** be equal to:
-  `K_KERNEL_STACK_SIZEOF(stack_obj)`, where `stack_obj `is defined as `K_KERNEL_STACK_DEFINE(stack_obj, STACKSIZE)`
- and we are only allowed to call `k_thread_create `with either of the two options as arguments, and
- these options are equivalent :)

Or, inverted:
K_KERNEL_STACK_SIZEOF(stack_obj) shall always return the `size` value used when defining stack_obj as `K_KERNEL_STACK_DEFINE(stack_obj, size)`

Do you agree?

Reply:
It's a little more complex than this.

For some invocation `K_THREAD_STACK_DEFINE(my_stack, X)`, the return value of `K_THREAD_STACK_SIZEOF(my_stack)` is *not* guaranteed to be `X`. They can be different values, they are not guaranteed to be the same. What will happen on some systems is that `K_THREAD_STACK_SIZEOF(my_stack)` is equal to 'X' *rounded up* to some alignment value.

However, `k_thread_create(my_thread, my_stack, Y, ...)` is guaranteed to work properly (and identically) if Y is either `X` or `K_THREAD_STACK_SIZEOF(my_stack)`.

What actually is equivalent, with some `K_THREAD_STACK_DEFINE(my_stack, X)`, is that `K_THREAD_STACK_SIZEOF(my_stack)` equals `Z_THREAD_STACK_SIZE_ADJUST(X)`. `setup_thread_stack()` in thread.c performs this calculation, which is how it is able to take either as the stack size value.

Reply:
> (and, btw, K_THREAD_STACK_SIZEOF is now renamed to K_KERNEL_STACK_SIZEOF.)

They are the same if user mode is not enabled, but differ slightly if user mode is enabled:

``
#define K_THREAD_STACK_SIZEOF(sym)	(sizeof(sym) - K_THREAD_STACK_RESERVED)
```
and 
```
#define K_KERNEL_STACK_SIZEOF(sym) (sizeof(sym) - K_KERNEL_STACK_RESERVED)
```

Key difference being that `K_THREAD_STACK_RESERVED` and `K_KERNEL_STACK_RESERVED` can differ (for example on x86 no memory for page tables or the privilege elevation stack is reserved for kernel-only stacks)